conn.go

```diff
@@ -62,36 +112,90 @@ func (c *conn) Close() error {
 }
 
 func (c *conn) reader(ctx context.Context) {
+	var discard uint
 	doneCh := ctx.Done()
 	for {
 		select {
 		case <-doneCh:
 			return
 		case mu := <-c.rCh:
 
+			c.readSeqL.Lock()
+
+			var skip bool
 			if err := mu.ctx.Err(); err != nil {
 				mu.errCh <- err
-				continue
+				skip = true
 			} else if deadline, ok := mu.ctx.Deadline(); ok {
 				if err := c.conn.SetReadDeadline(deadline); err != nil {
 					mu.errCh <- fmt.Errorf("setting read deadline to %v: %w", deadline, err)
-					continue
+					skip = true
 				}
 			} else if err := c.conn.SetReadDeadline(time.Time{}); err != nil {
 				mu.errCh <- fmt.Errorf("unsetting read deadline: %w", err)
+				skip = true
+			}
+
+			if !skip {
+				c.currReadSeq = mu.readSeq
+			}
+
+			c.readSeqL.Unlock()
+
+			if skip {
 				continue
 			}
 
-			err := resp3.Unmarshal(c.br, mu.unmarshalInto, c.rOpts)
+			var err error
+
+			// Discard messages queued up on the wire which aren't for this
+			// EncodeDecode call. Only discard messages if the EncodeDecode
+			// gives a value to marshal, as that indicates there's a specific
+			// response message in the queue which corresponds to it. If no
+			// value to marshal is given then the EncodeDecode will read in
+			// whatever the next message in the queue is.
+			if mu.marshal != nil {
+				for ; discard > 0; discard-- {
+					if err = resp3.Unmarshal(c.br, nil, c.rOpts); err != nil {
+						break
+					}
+				}
+			}
+
+			// if discarding didn't fail then read the actual desired response.
 			if err != nil {
-				// simplify things for the caller by translating network
-				// timeouts into DeadlineExceeded, since that's actually what
-				// happened.
-				var netErr net.Error
-				if errors.As(err, &netErr) && netErr.Timeout() {
-					err = context.DeadlineExceeded
+				err = resp3.Unmarshal(c.br, mu.unmarshalInto, c.rOpts)
```

Comment:
I can't comment the line above since it was not changed (thanks Github...) but the check should be `if err == nil` and not `if err != nil`

Reply:
ah yeah, good catch :bow: